blog/2020-12-04-nested-datatypes/assets/Nested.lhs

```diff
@@ -0,0 +1,638 @@
+could get a linear time conversion. However, it's probably not worth it in
+practice to do so.
+
+
+### Other operations
+
+Of course, these are not the only operations that one may define using perfect
+trees. However, these operations are good exemplars of what the set of
+operations might look like. Furthermore, the `Foldable` and `Traversable`
+instances mean that many operations can be written generically, for all
+trees.
+
+On the other hand, there is one significant difference between nested
+datatypes and GADTs: the former is supported directly by `GHC.Generics` but
+the latter is not. This means that the `Generic` class, from `GHC.Generics`
+is automatically derivable for nested datatypes, but not for GADTs. There may
+be a way to create an instance of the `Generic` type class by hand, but GHC
+cannot do so automatically.
+
+
+## Conclusions
+
+This is about as far as we can go with the comparison just using perfect
+ trees. There isn't all that much you can do with them. From my point of view,
+ I find the indexed version of the data structure a bit easier to understand
+ because of the contortions required for operations such as
+ `invertNTree`. However, maybe that is because I am already familiar with the
+ patterns of Dependent Haskell. If you are the opposite, perhaps this
+ explanation will serve as a Rosetta stone.
+
+### Other ways to define perfect trees
+
+There are still more ways to represent perfect trees in Haskell, using type
+ families, a finally-tagless encoding, LiquidHaskell, etc. A few of these
+ alternatives are collected
+ [here](https://github.com/sweirich/dth/tree/master/examples/perfect-trees).
+
+### Other nested datatypes
+
+Perfect trees are an example of a fairly constrained, symmetric and artificial
+data structure. So was it just a fluke that we could define a GADT analogue
+to the nested datatype definition?
+
+I don't think so.
+
+I know of at least two other examples of constrained data structures that can
+ be implemented by either nested datatypes and GADT-based alternatives.
+
+* [Well-scoped expressions](http://www.staff.city.ac.uk/~ross/papers/debruijn.html)
+
+A famous use of nested datatypes is to ensure that lambda calculus expressions
+ are well-scoped. This idea underlies the design of Kmett's
+ [bound](https://www.schoolofhaskell.com/user/edwardk/bound) library.
+
+However, instead of using a nested datatype, it is also possible to use a
+type-level natural number to bound the scopes of bound variables, as shown
+in [this implementation](https://github.com/sweirich/lennart-lambda/blob/master/lib/DeBruijnScoped.lhs).
+
+* [Finger trees](http://www.staff.city.ac.uk/~ross/papers/FingerTree.html)
+
+Haskell's implementation of the
+[sequence](https://hackage.haskell.org/package/containers-0.6.4.1/docs/Data-Sequence.html)
+data structure is built on FingerTrees. In the module
+[DFinger.lhs](DFinger.lhs) I've sketched out a nat-indexed replacement to the
```

Comment:
This link is unbound. Should be to https://github.com/sweirich/dth/blob/master/nested-datatypes/DFinger.lhs